### PR TITLE
WIP: Test listener improvements #494

### PR DESCRIPTION
**Pull request by sksamuel**
This is a wip, but see what you think. There's a few things left to implement, so don't worry about that, this PR is so I can get early feedback.

Changed the signatures of some listener functions to be more powerful, including the status of the tests. This will  address #468 #440 #469

I've also included the concept of a Suite (which isn't implemented yet) but will address #458

I'm fairly happen with it, except for the naming of the callback functions. I can't think of good names for "before the very first spec is executed" and "just before each spec is executed" and similarly for the after versions of those. So the names I've used here and totally subject to change.

### CODE REVIEW

File: kotlintest-tests/kotlintest-tests-core/src/test/kotlin/com/sksamuel/kotlintest/TestTimeoutTest.kt [hunk 1]
```diff
@@ -0,0 +1,24 @@
+//package com.sksamuel.kotlintest
```
Comment (Kerooker) [inline kotlintest-tests/kotlintest-tests-core/src/test/kotlin/com/sksamuel/kotlintest/TestTimeoutTest.kt]: Should this really be commented?

**Comment by Kerooker** [comment]
Making listener functions more powerful is pretty interesting!

The Suite concept is somewhat interesting, but wouldn't another level of abstraction be too much? I'm afraid that new users will have too much to grasp at once. Or are these "advanced features" not very available unless you want it?

I liked the `beforeSuite` and `afterSuite` names. If it's clear what a Suite is, it will be clear what these methods do, and I really liked their names!

File: kotlintest-tests/kotlintest-tests-core/src/test/kotlin/com/sksamuel/kotlintest/TestTimeoutTest.kt (hunk 1 above)
Comment (sksamuel) [reply kotlintest-tests/kotlintest-tests-core/src/test/kotlin/com/sksamuel/kotlintest/TestTimeoutTest.kt]: I didn't mean to check it in.

**Comment by sksamuel** [comment]
You would only have to use Suite if you wanted, you have to go looking for it.

**Comment by sksamuel** [comment]
I've removed suites for now, I'll work that into 3.3

**Pull request closed without merging**